### PR TITLE
Kafka Streams helper operations

### PR DESCRIPTION
**Pull request by jeqo**
Additional operations to include in `KafkaStreamsTracing` API to enable easier adoption of tracing as part of a Kafka Streams topology by emulating most-common operations (e.g. `KStream#map`, `KStream#mapValues`, `KStream#foreach`, `KStream#peek`) with Stream Transformers and Processors that have tracing enabled. 

Changes:

`KafkaStreamTracing` operations added:

- `ProcessorSupplier<K, V> foreach(String name, ForeachAction<K, V> action)`
- `TransformerSupplier<K, V, KeyValue<K, V>> peek(String name, ForeachAction<K, V> action)`
- `TransformerSupplier<K, V, KeyValue<K, V>> mark(String name)`
- `TransformerSupplier<K, V, KeyValue<KR, VR>> map(String name,
      KeyValueMapper<K, V, KeyValue<KR, VR>> mapper)`
- `ValueTransformerWithKeySupplier<K, V, VR> mapValues(String name,
      KeyValueMapper<K, V, VR> mapper)`
- `ValueTransformerSupplier<V, VR> mapValues(String name, ValueMapper<V, VR> mapper)`

Ref #825

### CODE REVIEW

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 1]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
     return new TracingValueTransformerWithKeySupplier<>(this, name, valueTransformerWithKey);
   }
 
+  /**
+   * Create a foreach processor, similar to {@link KStream#foreach(ForeachAction)}, where its action
+   * will be traced.
```
Comment (adriancole) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   * will be recorded in a new span with the indicated name.
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 2]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+    return new TracingProcessorSupplier<>(this, name, new AbstractProcessor<K, V>() {
+      @Override public void process(K key, V value) {
+        action.apply(key, value);
+      }
+    });
+  }
+
+  /**
+   * Create a peek transformer, similar to {@link KStream#peek(ForeachAction)}, where its action
+   * will be traced.
```
Comment (adriancole) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   * will be recorded in a new span with the indicated name.
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 3]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+      @Override public KeyValue<K, V> transform(K key, V value) {
+        action.apply(key, value);
+        return KeyValue.pair(key, value);
+      }
+    });
+  }
+
+  /**
+   * Create a mark transformer, similar to {@link KStream#peek(ForeachAction)}, but no action is executed.
+   * Instead only a span is created to mark the beginning or end of a task.
```
Comment (adriancole) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]: this needs a paragraph to explain why you would do this, also what a "task" is

Note: usually, this is what an annotation would be used for, so is a little confusing at first unless you also explain what will happen between beginning and end in the trace tree. Ex. does this result in two annotations, or a span than encloses the mapper?

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 4]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+    return new TracingTransformerSupplier<>(this, name, new AbstractTracingTransformer<K, V, KeyValue<K, V>>() {
+      @Override public KeyValue<K, V> transform(K key, V value) {
+        return KeyValue.pair(key, value);
+      }
+    });
+  }
+
+  /**
+   * Create a map transformer, similar to {@link KStream#map(KeyValueMapper)}, where its mapper action
+   * will be traced.
```
Comment (adriancole) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   * will be recorded in a new span with the indicated name.
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 5]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+        new AbstractTracingTransformer<K, V, KeyValue<KR, VR>>() {
+          @Override public KeyValue<KR, VR> transform(K key, V value) {
+            return mapper.apply(key, value);
+          }
+        });
+  }
+
+  /**
+   * Create a peek transformer, similar to {@link KStream#mapValues(ValueMapperWithKey)}, where its mapper action
+   * will be traced.
```
Comment (adriancole) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   * will be recorded in a new span with the indicated name.
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 6]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+        new AbstractTracingValueTransformerWithKey<K, V, VR>() {
+          @Override public VR transform(K readOnlyKey, V value) {
+            return mapper.apply(readOnlyKey, value);
+          }
+        });
+  }
+
+  /**
+   * Create a peek transformer, similar to {@link KStream#mapValues(ValueMapper)}, where its mapper action
+   * will be traced.
```
Comment (adriancole) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   * will be recorded in a new span with the indicated name.
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 7]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+
+  /**
+   * Create a map transformer, similar to {@link KStream#map(KeyValueMapper)}, where its mapper action
+   * will be recorded in a new span with the indicated name.
+   *
+   * <p>Simple example using Kafka Streams DSL:
+   * <pre>{@code
+   * StreamsBuilder builder = new StreamsBuilder();
+   * builder.stream(inputTopic)
+   *        .transform(kafkaStraemsTracing.map("myMap", (k, v) -> ...)
```
Comment (ImFlog) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   *        .transform(kafkaStreamsTracing.map("myMap", (k, v) -> ...)
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 8]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+
+  /**
+   * Create a peek transformer, similar to {@link KStream#mapValues(ValueMapperWithKey)}, where its mapper action
+   * will be recorded in a new span with the indicated name.
+   *
+   * <p>Simple example using Kafka Streams DSL:
+   * <pre>{@code
+   * StreamsBuilder builder = new StreamsBuilder();
+   * builder.stream(inputTopic)
+   *        .transform(kafkaStraemsTracing.mapValues("myMapValues", (k, v) -> ...)
```
Comment (ImFlog) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   *        .transform(kafkaStreamsTracing.mapValues("myMapValues", (k, v) -> ...)
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 9]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+
+  /**
+   * Create a peek transformer, similar to {@link KStream#mapValues(ValueMapper)}, where its mapper action
+   * will be recorded in a new span with the indicated name.
+   *
+   * <p>Simple example using Kafka Streams DSL:
+   * <pre>{@code
+   * StreamsBuilder builder = new StreamsBuilder();
+   * builder.stream(inputTopic)
+   *        .transform(kafkaStraemsTracing.mapValues("myMapValues", v -> ...)
```
Comment (ImFlog) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]:
```suggestion
   *        .transform(kafkaStreamsTracing.mapValues("myMapValues", v -> ...)
```

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java [hunk 10]
```diff
@@ -121,6 +129,132 @@ public KafkaStreams kafkaStreams(Topology topology, Properties streamsConfig) {
...
+   * will be recorded in a new span with the indicated name.
+   *
+   * <p>Simple example using Kafka Streams DSL:
+   * <pre>{@code
+   * StreamsBuilder builder = new StreamsBuilder();
+   * builder.stream(inputTopic)
+   *        .process(kafkaStraemsTracing.foreach("myForeach", (k, v) -> ...);
+   * }</pre>
+   */
+  public <K, V> ProcessorSupplier<K, V> foreach(String name, ForeachAction<K, V> action) {
```
Comment (ImFlog) [inline instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]: Maybe this has been discussed already but I don't find the parameter name to be explicit enough. In IDE, we are not sure if it is the span name or something else.
Could we rename it to spanName for example ?

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java (hunk 3 above)
Comment (jeqo) [reply instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]: agree.

File: instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java (hunk 10 above)
Comment (jeqo) [reply instrumentation/kafka-streams/src/main/java/brave/kafka/streams/KafkaStreamsTracing.java]: Totally agree 👍 `spanName` LGTM.

**Comment by jeqo** [comment]
loving suggestions function :)

**Comment by adriancole** [comment]
Thanks as always @jeqo and @ImFlog!

**Pull request merged**